Pay per chunk

joshuef opened the pull request:
starting to simplify payments to a per chunk basis

A first step in #638## Description

<!-- reviewpad:summarize:start -->
### Summary generated by Reviewpad on 16 Aug 23 09:51 UTC
This pull request includes multiple file diffs with various changes, such as renaming types and fields, importing and removing modules, updating function signatures and return types, and modifying method implementations. These changes seem to be focused on improving code clarity, making updates for better functionality, and enhancing the payment transaction handling.

Please review the changes thoroughly to ensure they align with the intended goals and do not introduce any unintended issues or conflicts. Let me know if you need further details or assistance with the code review.
<!-- reviewpad:summarize:end -->

joshuef committed:
feat!(protocol): remove chunk merkletree to simplify payment

starting to simplify payments to a per chunk basis

A first step in #638

joshuef committed:
feat!(protocol): get price and pay for each chunk individually

maqi commented on sn_client/src/api.rs:
```diff
@@ -292,7 +275,7 @@ impl Client {
         &self,
         chunk: Chunk,
-        payment: PaymentProof,
+        payment: Vec<DbcId>,
```
I thought there is only one `DbcId` of the payment for one chunk ?

joshuef replied:
I think you may need to reissue several DBCs sometimes if you have a lot of dust?

maqi commented on sn_client/src/wallet.rs:
```diff
@@ -104,49 +94,67 @@ impl WalletClient {
             .collect();
 
-        let number_of_records_to_pay = addrs_to_pay.len() as u64;
+        let _number_of_records_to_pay = addrs_to_pay.len() as u64;
```
this line can be removed

maqi commented on sn_client/src/wallet.rs:
```diff
@@ -104,49 +94,67 @@ impl WalletClient {
-        );
+        for content_addr in addrs_to_pay {
+            // TODO: parallelise this
+            let (payments, cost) = self
```
rename `payments` to `payment` as this is only about one chunk?

also, we don't care about the `cost`, unless for some logging purpose ?

joshuef replied:
could be many payments for one record, i think

maqi replied:
right, yeah may needed for `pay for each holder`, make sense

maqi commented on sn_client/src/wallet.rs:
```diff
@@ -104,49 +94,67 @@ impl WalletClient {
+        }
 
-        trace!("Making payment for {num_of_addrs} addresses of {amount_to_pay:?} nano tokens.");
+        Ok((proofs, Some(total_cost)))
```
same as above, we don't care about the `total_cost`, unless for some logging purpose ?

joshuef replied:
I think for gauging the total cost of a file we may want this?

maqi commented on sn_client/src/wallet.rs:
```diff
@@ -104,49 +94,67 @@
+
+        trace!("Making payment for of {amount_to_pay:?} nano tokens to each node in close group at {content_addr:?}.");
+
+        // TODO: This needs to go out to each CLOSEGROUP of addresses
```
you mean paying each individual holder ?

hmm...  I don't think we need to contact each holder individually, 
just `get_closest(chunk_addr)` then transfer some amount to each fetched close_peer, and send the transfer to network shall be enough?

just, that means the payment_proof needs to hold 8 spend_ids, and on validation, the holder only needs to validate the spend to itself (or still ALL for security?)

joshuef replied:
Each CLOSE_GROUP for each record address needs to return a price, which should then be paid. 

on validation, each would just do itself, I think?

maqi replied:
> Each CLOSE_GROUP for each record address needs to return a price, which should then be paid.

ok. I thought a close_group can charge with the same.

> on validation, each would just do itself, I think?

yeah. but that's can be done on receiving the record via `put_record` , instead of contacting them directly individually.

joshuef commented:
This closes #638

joshuef committed:
fix(logging): get log name per bin

joshuef committed:
fix(client): use boostrap and fire Connecting event

Previously we were hanging waiting on inactivity before we continued, now we
push the loop forward and can us bootstrap to get us going

joshuef committed:
ci: fix client log naming

joshuef commented:
Closing as #649 is atop this and working better
